client/components/domains/registrant-extra-info/fr-form.jsx

```diff
@@ -0,0 +1,351 @@
+		// immediatly to to make sure there's something valid in the parent
+		this.props.onStateChange( values );
+	}
+
+	handleDobChangeEvent = ( event ) => {
+		// setState() is not syncronous :(
+		const newState = this.newStateFromEvent( event );
+		const { dob_years, dob_months, dob_days } = newState;
+
+		const dateOfBirth = ( dob_years && dob_months && dob_days ) &&
+			[ dob_years, dob_months, dob_days ].join( '-' );
+
+		if ( dateOfBirth ) {
+			debug( 'Setting dateOfBirth to ' + dateOfBirth +
+				( moment( dateOfBirth, 'YYYY-MM-DD' ).isValid() ? '' : ' (invalid)' ) );
+		}
+
+		this.props.onStateChange( {
+			...newState,
+			...( dateOfBirth ? { dateOfBirth } : {} )
+		} );
+	}
+
+	newStateFromEvent( event ) {
+		return {
+			...this.props.values,
+			[ event.target.name ]: event.target.value,
+		};
+	}
+
+	handleChangeEvent = ( event ) => {
+		this.props.onStateChange( this.newStateFromEvent( event ) );
+	}
+
+	// We need a deep comparison to check inside props.values
+	shouldComponentUpdate = ( nextProps ) => {
+		return ! isEqual( this.props, nextProps );
+	}
+
+	render = () => {
+		const translate = this.props.translate;
+		const {
+			registrantType
+		} = { ...emptyValues, ...this.props.values };
+
+		return (
+			<form>
+				<FormFieldset>
+					{ translate(
+						'We need some extra details to register domains ending in ".fr".'
+					) }
+				</FormFieldset>
+				<FormFieldset>
+					<FormLegend>
+						{ translate( "Who's this domain for?" ) }
+					</FormLegend>
+					<FormLabel>
+						<FormRadio value="individual"
+							name="registrantType"
+							checked={ 'individual' === registrantType }
+							onChange={ this.handleChangeEvent } />
+						<span>{ translate( 'An individual' ) }</span>
+					</FormLabel>
+
+					<FormLabel>
+						<FormRadio value="organization"
+							name="registrantType"
+							checked={ 'organization' === registrantType }
+							onChange={ this.handleChangeEvent } />
+						<span>{ translate( 'A company or organization' ) }</span>
+					</FormLabel>
+				</FormFieldset>
+
+				{ 'individual' === registrantType
+					? this.renderPersonalFields()
+					: this.renderOrganizationFields() }
+
+				{ this.props.children }
+			</form>
+		);
+	}
+
+	renderPersonalFields = () => {
+		const translate = this.props.translate;
+		const {
+			countryOfBirth,
+			dob_days,
+			dob_months,
+			dob_years,
+			placeOfBirth,
+			postalCodeOfBirth,
+		} = { ...emptyValues, ...this.props.values };
+
+		return (
+			<div>
+				<FormFieldset>
+					<FormLabel>
+						{ translate( 'Country of Birth' ) }
+					</FormLabel>
+					<FormCountrySelect
+						name="countryOfBirth"
+						value={ countryOfBirth }
+						countriesList={ this.props.countriesList }
+						onChange={ this.handleChangeEvent } />
+				</FormFieldset>
+
+				<FormFieldset>
+					<FormLegend>
+						{ translate( 'Date of Birth' ) }
+					</FormLegend>
+					<div className="registrant-extra-info__dob-inputs">
```

Comment:
How about using https://wpcalypso.wordpress.com/devdocs/design/date-picker instead? It should provide a much better UX (not to mention that for a person in US, the month should be first in the input, etc. ;)) with hopefully less code 😃 
Otherwise, this approach needs more strict validation (you can input more than 4 digits for the year, etc.) - all that should be taken care of by DatePicker.